src/audio/drc/drc_generic.c

```diff
@@ -0,0 +1,453 @@
+{
+	const float kA = Q_CONVERT_QTOF(p->kA, 12);
+	const float kB = Q_CONVERT_QTOF(p->kB, 12);
+	const float kC = Q_CONVERT_QTOF(p->kC, 12);
+	const float kD = Q_CONVERT_QTOF(p->kD, 12);
+	const float kE = Q_CONVERT_QTOF(p->kE, 12);
+	const float attack_frames = Q_CONVERT_QTOF(p->attack_frames, 20);
+
+	/* Calculate desired gain */
+	float desired_gain = Q_CONVERT_QTOF(state->detector_average, 30);
+
+	/* Pre-warp so we get desired_gain after sin() warp below. */
+	float scaled_desired_gain = warp_asinf(desired_gain);
+
+	/* Deal with envelopes */
+
+	/* envelope_rate is the rate we slew from current compressor level to
+	 * the desired level.  The exact rate depends on if we're attacking or
+	 * releasing and by how much.
+	 */
+	float envelope_rate;
+
+	int is_releasing = scaled_desired_gain > Q_CONVERT_QTOF(state->compressor_gain, 30);
+
+	/* compression_diff_db is the difference between current compression
+	 * level and the desired level. */
+	float compression_diff_db = linear_to_decibels(
+		Q_CONVERT_QTOF(state->compressor_gain, 30) / scaled_desired_gain);
+
+	if (is_releasing) {
+		/* Release mode - compression_diff_db should be negative dB */
+		state->max_attack_compression_diff_db = INT32_MIN;
+
+		/* Fix gremlins. */
+		if (isbadf(compression_diff_db))
+			compression_diff_db = -1;
+
+		/* Adaptive release - higher compression (lower
+		 * compression_diff_db) releases faster. Contain within range:
+		 * -12 -> 0 then scale to go from 0 -> 3
+		 */
+		float x = compression_diff_db;
+		x = max(-12.0f, x);
+		x = min(0.0f, x);
+		x = 0.25f * (x + 12);
+
+		/* Compute adaptive release curve using 4th order polynomial.
+		 * Normal values for the polynomial coefficients would create a
+		 * monotonically increasing function.
+		 */
+		float x2 = x * x;
+		float x3 = x2 * x;
+		float x4 = x2 * x2;
+		float release_frames = kA + kB * x + kC * x2 + kD * x3 + kE * x4;
+
+#define kSpacingDb 5
+		float db_per_frame = kSpacingDb / release_frames;
+		envelope_rate = decibels_to_linear(db_per_frame);
+	} else {
+		/* Attack mode - compression_diff_db should be positive dB */
+
+		/* Fix gremlins. */
+		if (isbadf(compression_diff_db))
+			compression_diff_db = 1;
+
+		/* As long as we're still in attack mode, use a rate based off
+		 * the largest compression_diff_db we've encountered so far.
+		 */
+		state->max_attack_compression_diff_db =
+			max(state->max_attack_compression_diff_db,
+			    Q_CONVERT_FLOAT(compression_diff_db, 24));
+
+		float eff_atten_diff_db =
+			max(0.5f, Q_CONVERT_QTOF(state->max_attack_compression_diff_db, 24));
+
+		float x = 0.25f / eff_atten_diff_db;
+		envelope_rate = 1 - powf(x, 1 / attack_frames);
+	}
+
+	state->envelope_rate = Q_CONVERT_FLOAT(envelope_rate, 30);
+	state->scaled_desired_gain = Q_CONVERT_FLOAT(scaled_desired_gain, 30);
+}
+
+/* Convert an audio sample from floating point format to s16 format and saturate to [-1.0, 1.0].*/
+static int16_t drc_float2s16(float f)
```

Comment:
is this temp?

Reply:
Yes, this is temp. We should replace all float usage.